### DISCUSSION TITLE
Editable datagrid?

### DISCUSSION THREAD

**Discussion by golddragon007** [AvaloniaUI/Avalonia#8210] [Q&A] [answered]
Hello,

I am looking for the DataGrid control, but I can't figure out why I can't allow the user to edit its values and/or content. Does this behave differently from the WPF component?
view.axaml:
```
                <DataGrid Grid.Row="1" CanUserResizeColumns="True" Items="{Binding Topics, Mode=TwoWay}">
			<DataGrid.Columns>
				<DataGridTextColumn Header="Topics" Binding="{Binding}" IsReadOnly="False" />
			</DataGrid.Columns>
		</DataGrid>
```
viewviewmodel.cs:
```
        private List<string> _topics;
        public List<string> Topics
        {
            get
            {
                return _topics;
            }
            set
            {
                _topics = value;
                this.RaisePropertyChanged(nameof(Topics));
            }
        }
```
app.axaml:
```
        <Application.Styles>
        <FluentTheme Mode="Light"/>
		<StyleInclude Source="avares://Avalonia.Controls.DataGrid/Themes/Fluent.xaml"/>
	</Application.Styles>
```

End result:
![image](https://user-images.githubusercontent.com/8385152/170683431-7b9e895a-a7f4-4b2d-9394-b61c54aab5a0.png)

Why it does not save the edited cell changes?
How to add a new item?
How to allow users to reorder rows?
How to delete an already existing row?

**Comment by maxkatz6** [comment, marked as answer]
Hi!

> Does this behave differently from the WPF component?

It's silverlight port, but differs from wpf one slightly. 

>  Why it does not save the edited cell changes?

String is immutable object. Use proper models with property changed notifications as an item. 

> How to add a new item?
How to allow users to reorder rows?
How to delete an already existing row?

Modify source collection. Make it a ObservableCollection or any other type that implements INotifyCollectionChanged

**Reply by golddragon007** [reply]
Ok, Well the editing works now.
And then I suppose I need to implement the rest manually, as it has no such support for adding, reordering (not by header sorting) and deleting items. At least I did not find anything about those.

**Reply by maxkatz6** [reply]
WPF DataGrid doesn't have row reordering either, you need to implement dragndrop there manually. And from what I remember it doesn't have magic "add" and "remove" buttons either, or does it?
Third party rich alternatives have it though.

**Reply by golddragon007** [reply]
Yeah, -I guess- for sure the reordering is not implemented in the native (as I've found now this: http://www.hardcodet.net/2009/03/moving-data-grid-rows-using-drag-and-drop), but the add and delete are:
"Set the [CanUserAddRows](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.datagrid.canuseraddrows?view=windowsdesktop-6.0) and [CanUserDeleteRows](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.datagrid.canuserdeleterows?view=windowsdesktop-6.0) properties to specify whether a user can add or delete rows. A user can delete selected rows by pressing the DELETE key."
- https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.datagrid?view=windowsdesktop-6.0

Well, anyway, you confirmed that I'm not blind to not finding it, as it does not exist here. :D So I have to deal with it.
